File: sdks/kotlin/src/test/kotlin/dev/wheso/TraceReceiverTest.kt
// 受信ノードのトレースベクタの照合（Kotlin、層 2: 決定同一）。
//
// 凍結トレース（spec/vectors/trace-receiver.jsonl）を Kotlin の判断コアへ流し、
// 出力コマンド列が TypeScript の参照実装と**完全に一致**することを確かめる。
// 相違した場合はベクタではなく実装を直す（ADR-0012）。
package dev.wheso

import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

class TraceReceiverTest {
    /** 生成器（tools/traces-receiver.ts）と一致させる初期予算。 */
    private val initialBudgetBytesPerSec: Long = 7_000_000L

    private fun longOf(obj: JsonObject, key: String): Long {
        val value = obj[key]?.jsonPrimitive?.content
        return value?.toLongOrNull() ?: error("$key が整数ではない")
    }

    private fun toEvent(input: JsonObject): ReceiverEvent {
        return when (input["kind"]?.jsonPrimitive?.content) {
            "subscribe" -> {
                val raw = input["entries"]?.jsonArray ?: error("entries が無い")
                ReceiverEvent.SubscribeList(
                    raw.map { element ->
                        val entry = element.jsonObject
                        SubscribeEntry(
                            senderId = longOf(entry, "senderId"),
                            channel = longOf(entry, "channel"),
                            maxSpatialId = longOf(entry, "maxSpatialId"),
                            maxTemporalId = longOf(entry, "maxTemporalId"),
                        )
                    },
                )
            }
            "leave" -> ReceiverEvent.Leave(longOf(input, "id"))
            "visibility" -> ReceiverEvent.Visibility(
                input["visible"]?.jsonPrimitive?.content == "true",
            )
            "budget" -> ReceiverEvent.Budget(longOf(input, "bytesPerSec"))
            "activeSpeaker" -> {
                // null は「発話者なし」を意味する。欄の欠落と区別する。
                val value = input["activeSpeaker"] ?: input["id"] ?: error("id が無い")
                if (value is JsonNull) {
                    ReceiverEvent.ActiveSpeaker(null)
                } else {
                    ReceiverEvent.ActiveSpeaker(value.jsonPrimitive.content.toLongOrNull())
                }
            }
            "displaySize" -> ReceiverEvent.DisplaySize(
                senderId = longOf(input, "senderId"),
                channel = longOf(input, "channel"),
                width = longOf(input, "width"),
            )
            "report" -> {
                val samples = input["delayUs"]?.jsonArray ?: error("delayUs が無い")
                ReceiverEvent.Report(
                    samples.map { it.jsonPrimitive.content.toLongOrNull() ?: error("整数でない") },
                )
            }
            "media" -> ReceiverEvent.Media(
                from = longOf(input, "from"),
                ch = longOf(input, "ch"),
                sid = longOf(input, "sid"),
                tid = longOf(input, "tid"),
                seq = input["seq"]?.jsonPrimitive?.content?.toLongOrNull() ?: 0L,
            )
            "timer" -> ReceiverEvent.Timer
            else -> error("未知のイベント: ${input["kind"]}")
        }
    }

    /** 出力コマンドを TypeScript と同じ JSON 表現へ写す。欄名も一致させる。 */
    private fun toJson(command: ReceiverCommand): JsonObject = when (command) {
        is ReceiverCommand.SubscribeChange -> buildJsonObject {
            put("kind", JsonPrimitive("subscribeChange"))
            put("to", JsonPrimitive(command.to))
            put("channel", JsonPrimitive(command.channel))
            put("want", JsonPrimitive(command.want))
            put("maxSpatialId", JsonPrimitive(command.maxSpatialId))
            put("maxTemporalId", JsonPrimitive(command.maxTemporalId))
        }
        is ReceiverCommand.KeyframeRequest -> buildJsonObject {
            put("kind", JsonPrimitive("keyframeRequest"))
            put("for", JsonPrimitive(command.targetId))
            put("channel", JsonPrimitive(command.channel))
            put("spatialId", JsonPrimitive(command.spatialId))
        }
        is ReceiverCommand.SetTier -> buildJsonObject {
            put("kind", JsonPrimitive("setTier"))
            put("for", JsonPrimitive(command.targetId))
            put("channel", JsonPrimitive(command.channel))
            put("tier", JsonPrimitive(command.tier))
        }
        is ReceiverCommand.Forward -> buildJsonObject {
            put("kind", JsonPrimitive("forward"))
            put("to", buildJsonArray { command.to.forEach { add(JsonPrimitive(it)) } })
        }
        is ReceiverCommand.Drop -> buildJsonObject {
            put("kind", JsonPrimitive("drop"))
            put("priority", JsonPrimitive(command.priority))
            put("count", JsonPrimitive(command.count))
        }
        is ReceiverCommand.Notify -> buildJsonObject {
            put("kind", JsonPrimitive("notify"))
            put("code", JsonPrimitive(command.code))
        }
        is ReceiverCommand.Ack -> buildJsonObject {
            put("kind", JsonPrimitive("ack"))
            put("senderId", JsonPrimitive(command.senderId))
            put("channel", JsonPrimitive(command.channel))
            put("spatialId", JsonPrimitive(command.spatialId))
            put("highestSeq", JsonPrimitive(command.highestSeq))
        }
    }

    /** 欄の順序に依存せず、欄の過不足も検出する形へ正規化する。 */
    private fun normalize(array: JsonArray): List<Map<String, String>> =
        array.map { element ->
            element.jsonObject.entries.associate { (key, value) ->
                key to (if (value is JsonArray) value.toString() else value.jsonPrimitive.content)
            }
        }

    @Test
    fun frozenReceiverTraceMatchesTypescriptReference() {
        val lines = File("../../spec/vectors/trace-receiver.jsonl")
            .readLines()
            .filter { it.isNotBlank() }
        assertTrue(lines.size > 100, "トレースが十分な行数を持つ")

        val header = Json.parseToJsonElement(lines.first()).jsonObject
        assertEquals("receiver", header["unit"]?.jsonPrimitive?.content, "受信ノードのトレースである")

        var state = initialReceiverState(initialBudgetBytesPerSec)
        var pending: JsonObject? = null
        var checked = 0

        for (line in lines.drop(1)) {
            val row = Json.parseToJsonElement(line).jsonObject
            val input = row["in"]
            if (input != null) {
                pending = input.jsonObject
                continue
            }
            val out = row["out"] ?: continue
            val event = pending ?: error("出力に対応する入力が無い")
            pending = null
            val result = receiverStep(state, toEvent(event))
            state = result.state
            val actual = buildJsonArray { result.commands.forEach { add(toJson(it)) } }
            assertEquals(
                normalize(out.jsonArray),
                normalize(actual),
                "入力 $event に対する出力が一致する",
            )
            checked += 1
        }
        assertTrue(checked > 100, "十分な行数を照合した（実際 $checked）")
    }
}

File: sdks/dart/test/trace_receiver_test.dart
// 受信ノードのトレースベクタの照合（Dart、層 2: 決定同一）。
//
// 凍結トレース（spec/vectors/trace-receiver.jsonl）を Dart の判断コアへ流し、
// 出力コマンド列が TypeScript の参照実装と**完全に一致**することを確かめる。
// 相違した場合はベクタではなく実装を直す（ADR-0012）。

import 'dart:convert';
import 'dart:io';

import 'package:test/test.dart';
import 'package:wheso_client/src/receiver_core.dart';

Map<String, Object?> readMap(Object? value, String where) {
  if (value is Map<String, Object?>) {
    return value;
  }
  throw StateError('$where: 連想配列ではない');
}

List<Object?> readList(Object? value, String where) {
  if (value is List<Object?>) {
    return value;
  }
  throw StateError('$where: 配列ではない');
}

int readInt(Map<String, Object?> map, String key, String where) {
  final value = map[key];
  if (value is int) {
    return value;
  }
  throw StateError('$where: $key が整数ではない');
}

ReceiverEvent toEvent(Map<String, Object?> input) {
  final kind = input['kind'];
  switch (kind) {
    case 'subscribe':
      final entries = <SubscribeEntry>[];
      for (final raw in readList(input['entries'], 'subscribe entries')) {
        final entry = readMap(raw, 'subscribe entry');
        entries.add(SubscribeEntry(
          senderId: readInt(entry, 'senderId', 'entry'),
          channel: readInt(entry, 'channel', 'entry'),
          maxSpatialId: readInt(entry, 'maxSpatialId', 'entry'),
          maxTemporalId: readInt(entry, 'maxTemporalId', 'entry'),
        ));
      }
      return SubscribeListEvent(entries: entries);
    case 'leave':
      return LeaveEvent(id: readInt(input, 'id', 'leave'));
    case 'visibility':
      final visible = input['visible'];
      if (visible is! bool) {
        throw StateError('visibility: visible が真偽値ではない');
      }
      return VisibilityEvent(visible: visible);
    case 'budget':
      return BudgetEvent(bytesPerSec: readInt(input, 'bytesPerSec', 'budget'));
    case 'goodput':
      return GoodputEvent(bytesPerSec: readInt(input, 'bytesPerSec', 'goodput'));
    case 'activeSpeaker':
      // null は「発話者なし」を意味する。欄の欠落と区別する。
      if (!input.containsKey('id')) {
        throw StateError('activeSpeaker: id が無い');
      }
      final id = input['id'];
      if (id == null) {
        return const ActiveSpeakerEvent(id: null);
      }
      if (id is! int) {
        throw StateError('activeSpeaker: id が整数でも null でもない');
      }
      return ActiveSpeakerEvent(id: id);
    case 'displaySize':
      return DisplaySizeEvent(
        senderId: readInt(input, 'senderId', 'displaySize'),
        channel: readInt(input, 'channel', 'displaySize'),
        width: readInt(input, 'width', 'displaySize'),
      );
    case 'catalog':
      final rawEntries = readList(input['entries'], 'catalog entries');
      final catalogEntries = <CatalogLadder>[];
      for (final rawEntry in rawEntries) {
        final entry = readMap(rawEntry, 'catalog entry');
        final rawRungs = readList(entry['rungs'], 'catalog rungs');
        final rungs = <CatalogRung>[];
        for (final rawRung in rawRungs) {
          final r = readMap(rawRung, 'catalog rung');
          rungs.add(CatalogRung(
            sid: readInt(r, 'sid', 'rung'),
            width: readInt(r, 'width', 'rung'),
            height: readInt(r, 'height', 'rung'),
            framerate: readInt(r, 'framerate', 'rung'),
            temporalLayers: readInt(r, 'temporalLayers', 'rung'),
            targetBitrate: readInt(r, 'targetBitrate', 'rung'),
          ));
        }
        catalogEntries.add(CatalogLadder(
          senderId: readInt(entry, 'senderId', 'catalog entry'),
          channel: readInt(entry, 'channel', 'catalog entry'),
          rungs: rungs,
        ));
      }
      return CatalogEvent(entries: catalogEntries);
    case 'report':
      final delayUs = <int>[];
      for (final sample in readList(input['delayUs'], 'report delayUs')) {
        if (sample is! int) {
          throw StateError('report delayUs: 整数ではない');
        }
        delayUs.add(sample);
      }
      return ReportEvent(delayUs: delayUs);
    case 'media':
      final seq = input['seq'];
      return MediaEvent(
        from: readInt(input, 'from', 'media'),
        ch: readInt(input, 'ch', 'media'),
        sid: readInt(input, 'sid', 'media'),
        tid: readInt(input, 'tid', 'media'),
        seq: seq is int ? seq : 0,
      );
    case 'keyframeRequest':
      return KeyframeRequestEvent(
        senderId: readInt(input, 'senderId', 'keyframeRequest'),
        channel: readInt(input, 'channel', 'keyframeRequest'),
        spatialId: readInt(input, 'spatialId', 'keyframeRequest'),
      );
    case 'timer':
      return const TimerEvent();
  }
  throw StateError('未知のイベント: $kind');
}

/// 出力コマンドを TypeScript と同じ JSON 表現へ写す。欄名も順序も一致させる。
Map<String, Object?> toJson(ReceiverCommand command) {
  switch (command) {
    case SubscribeChangeCommand():
      return <String, Object?>{
        'kind': 'subscribeChange',
        'to': command.to,
        'channel': command.channel,
        'want': command.want,
        'maxSpatialId': command.maxSpatialId,
        'maxTemporalId': command.maxTemporalId,
      };
    case KeyframeRequestCommand():
      return <String, Object?>{
        'kind': 'keyframeRequest',
        'for': command.targetId,
        'channel': command.channel,
        'spatialId': command.spatialId,
      };
    case SetTierCommand():
      return <String, Object?>{
        'kind': 'setTier',
        'for': command.targetId,
        'channel': command.channel,
        'tier': command.tier,
      };
    case ForwardCommand():
      return <String, Object?>{'kind': 'forward', 'to': command.to};
    case DropCommand():
      return <String, Object?>{'kind': 'drop', 'priority': command.priority, 'count': command.count};
    case NotifyCommand():
      return <String, Object?>{'kind': 'notify', 'code': command.code};
    case AckCommand():
      return <String, Object?>{
        'kind': 'ack',
        'senderId': command.senderId,
        'channel': command.channel,
        'spatialId': command.spatialId,
        'highestSeq': command.highestSeq,
      };
  }
}

void main() {
  test('受信ノードの凍結トレースが参照実装と完全一致する', () {
    final file = File('../../spec/vectors/trace-receiver.jsonl');
    final lines = file.readAsLinesSync().where((line) => line.trim().isNotEmpty).toList();
    expect(lines.length, greaterThan(100));

    final header = readMap(jsonDecode(lines.first), 'header');
    expect(header['unit'], equals('receiver'));

    var state = initialReceiverState();
    Map<String, Object?>? pending;
    // 入力行の時刻。AIMD の待ち（RATE_HOLD_MS）の判定に使う。
    int pendingT = 0;
    var checked = 0;

    for (final line in lines.skip(1)) {
      final row = readMap(jsonDecode(line), 'row');
      final input = row['in'];
      if (input != null) {
        pending = readMap(input, 'in');
        final rawT = row['t'];
        pendingT = rawT is int ? rawT : 0;
        continue;
      }
      final out = row['out'];
      if (out == null) {
        continue;
      }
      final expected = readList(out, 'out');
      final event = pending;
      if (event == null) {
        throw StateError('出力に対応する入力が無い');
      }
      pending = null;
      final result = receiverStep(state, toEvent(event), pendingT);
      state = result.state;
      final actual = result.commands.map(toJson).toList();
      expect(jsonEncode(actual), equals(jsonEncode(expected)),
          reason: '入力 ${jsonEncode(event)} に対する出力が一致する');
      checked += 1;
    }
    expect(checked, greaterThan(100), reason: '十分な行数を照合した');
  });
}
